go/database/mpt/proof_test.go

```diff
@@ -401,6 +402,139 @@ func TestWitnessProof_Extract_EmbeddedNode_In_Proof(t *testing.T) {
+			got, exist, err := proof.GetBalance(rootHash, address)
+			if !exist || err != nil {
+				return exist, err
+			}
+			if got, want := got, balance; got != want {
+				t.Errorf("unexpected balance: got %v, want %v", got, want)
+			}
+			return exist, err
+		}},
+		"GetCodeHash": {test: func(t *testing.T, proof WitnessProof, rootHash common.Hash, address common.Address, key common.Key) (bool, error) {
+			got, exist, err := proof.GetCodeHash(rootHash, address)
+			if !exist || err != nil {
+				return exist, err
+			}
+			if got, want := got, hash; got != want {
+				t.Errorf("unexpected code hash: got %v, want %v", got, want)
+			}
+			return exist, err
+		}},
+		"GetValue": {test: func(t *testing.T, proof WitnessProof, rootHash common.Hash, _ common.Address, key common.Key) (bool, error) {
+			got, exist, err := proof.GetState(rootHash, address, key)
+			if !exist || err != nil {
+				return exist, err
+			}
+			if got, want := got, value; got != want {
+				t.Errorf("unexpected value: got %v, want %v", got, want)
+			}
+			return exist, err
+		}},
+	}
+
+	desc := &Extension{
+		path: AddressToNibblePath(address, ctxt)[0:30],
+		next: &Account{address: address, pathLength: 34, info: info,
+			storage: &Extension{
+				path: KeyToNibblePath(key, ctxt)[0:40],
+				next: &Value{key: key, length: 24, value: value},
+			}},
+	}
+
+	root, node := ctxt.Build(desc)
+	totalProof := createReferenceProof(t, ctxt, &root, node)
+
+	handle := node.GetReadHandle()
+	defer handle.Release()
+
+	rootHash, dirty := handle.Get().GetHash()
+	if dirty {
+		t.Fatalf("expected node to be clean")
+	}
+
+	// corrupt non-root node in the proof
+	corruptedProof := WitnessProof{maps.Clone(totalProof.proofDb)}
+	for k := range corruptedProof.proofDb {
+		if k != rootHash {
+			corruptedProof.proofDb[k] = []byte{0xAA, 0xBB, 0xCC, 0xDD}
+		}
+	}
+
+	for name, test := range tests {
+		t.Run(fmt.Sprintf("%s exists", name), func(t *testing.T) {
+			exists, err := test.test(t, totalProof, rootHash, address, key)
+			if err != nil {
+				t.Fatalf("unexpected error: %v", err)
+			}
+			if !exists {
+				t.Fatalf("proof should exist")
+			}
+		})
+		t.Run(fmt.Sprintf("%s do not exists", name), func(t *testing.T) {
+			exists, err := test.test(t, totalProof, rootHash, common.Address{}, common.Key{})
+			if err != nil {
+				t.Fatalf("unexpected error: %v", err)
+			}
+			if exists {
+				t.Fatalf("proof should not exist")
+			}
+		})
+		t.Run(fmt.Sprintf("%s invalid", name), func(t *testing.T) {
+			_, err := test.test(t, corruptedProof, rootHash, address, key)
+			if err == nil {
+				t.Fatalf("expected error")
+			}
+		})
+		t.Run(fmt.Sprintf("%s rootHash does not exists", name), func(t *testing.T) {
+			exists, err := test.test(t, totalProof, common.Hash{}, address, key)
+			if err != nil {
+				t.Fatalf("unexpected error: %v", err)
+			}
+			if exists {
+				t.Fatalf("proof should not exist")
+			}
+		})
+	}
+}
+
```

Comment:
I think this test should pass, but it fails right now:
```
func TestWitnessProof_Proof_Zero_Nonces(t *testing.T) {
	ctrl := gomock.NewController(t)
	ctxt := newNodeContextWithConfig(t, ctrl, S5LiveConfig)

	root, node := ctxt.Build(&Account{
		pathLength: 64,
		info:       AccountInfo{Balance: common.Balance{1}},
	})

	rootHash, _ := ctxt.getHashFor(&root)
	totalProof := createReferenceProof(t, ctxt, &root, node)
	for _, addr := range []common.Address{{0}, {1}} { // one non-empty account and an empty account
		t.Run(fmt.Sprintf("account %v", addr), func(t *testing.T) {
			nonce, found, err := totalProof.GetNonce(rootHash, addr)
			if err != nil {
				t.Fatalf("unexpected error extracting proof: %v", err)
			}
			if !found {
				t.Errorf("proof should cover account")
			}
			if got, want := nonce, (common.Nonce{0}); got != want {
				t.Errorf("unexpected nonce: got %v, want %v", got, want)
			}
		})
	}
}
```

Can you have a look?

Reply:
Thanks for this test case. It helped to uncover one more issue comparing address/key converted to path. In some cases the address or key was padded with zero, while this zero is not included in the iterated path. I fixed it by comparing nibbles to nibbles, not Path instances.

Furthermore, the test did not pass because at the moment there is limitation of address size to 40 nibbles, while this test needed 64 nibbles.  It is to be fixed before release. 

Finally, I modified the test and impl. and at the moment it passes for the address `0` . It still fails for the address `1`, but here I do not understand why it should pass? This account is not in the proof and for this reason `false` is returned.

Reply:
It should pass for the account `1` because all the information needed to proof that its nonce is `0` is in the proof.

The argument goes as follows:
 - The proof contains a root node which encodes the account information for address `0`. 
 - The fact that this is the root node guarantees that there is no account node in the MPT covering account `1`.  
 - The default nonce value of every account that is not in an MPT is 0;

Thus, in this case, the proof for account `0` also covers the information of all other possible accounts.

This example is just the smallest possible example of a proof for missing information. You can also argue it the other way around:
- create an MPT with a single account `0` with a positive balance
- ask the MPT for a proof about the nonce of account `1`
- the resulting proof should be equivalent to the proof for the nonce of account `0`, which is what is extracted in the test.

Reply:
thanks for explanation. I think the point of misunderstanding was that I thought that the resulting `bool` mens if the requested account/slot is in the proof. But it should represent if the proof is complete, or not, right.

In practices, should it work that everytime you ask for an account or storage, and it can find any `AccountNode`, `ValueNode`, or `EmptyNode`, it returns true.  Otherwise - `Branch`/`Extension` Node it returns false?

I guess it should work the same for the method that `Extract`s the proof - for consistency. Right?

Reply:
I implemented it in this commit: b948266c739b9814166414735a4d66d503824300 

please have a look if it is what you meant

Reply:
The traversal through the proof may also end in an `Extension` node and data can be derived from that.

For instance, assume you have two accounts A and B such that their hashes have the first 2 bytes in common -- this is a chance of 1 : 65536, so not that unlikely. A trie with those two nodes would look like this:
```
Extension{
  path: {a, b}
  next: Branch{
     c: Account{ address: A },
     d: Account{ address: B },
  }
}
```
For this, if I asked for a proof for an account C, where C's hash starts with an `aa`, the navigation would get me to the extension node and then it should terminate. The proof only contains the extension node.

When then I ask the proof for C for C's nonce, the algorithm would navigate through the proof and stop in the Extension node. There, it concludes that since the extension-nodes path is `ab` there is no account with a prefix `aa` in the trie, and thus that C`s state is empty. Consequently, C`s nonce is `0`.

To create a test case for this, you might want to use the two keys from [here](https://github.com/Fantom-foundation/Carmen/blob/ad1657431037430a60f7456088947e5bfd16972a/go/database/mpt/verification_test.go#L577-L584). Back, when I created it, I did some brute-forcing to find a key pair with a long common prefix, and this was the result.

Reply:
Also, thinking about it, a `Branch` node can also never be the terminal node.

The only time when you should return `false` is if you do a lookup in the proof map and there is a node missing. Simple as that.

Reply:
good point, I got it. It will simplify the implementation a lot